### PR TITLE
LagESPModule (very useful)

### PR DESCRIPTION
**Pull request by oragejuice**
very very useful (you should accept now)

### CODE REVIEW

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+            /* register that a rubberband happened */
+
+            list.add(new RubberBand(
+                    mc.player.getPositionVector(),
+                    new Vec3d(p.getX(), p.getY(), p.getZ()
+                    )));
+
+        }
+
+
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:55]: why so many blank lines 😭

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        super("LagESP", Category.VISUAL, "Shows a path from where you got rubber banded");
+    }
+
+    public static Setting<Float> fadeSpeed = new Setting<>("FadeSpeed", 1F, 2F, 10F, 1)
+            .setDescription("How long the rubber band should show for");
+
+    public static Setting<Float> width = new Setting<>("Width", 0.1F, 2F, 5F, 1)
+            .setDescription("The width of the lines");
+
+    public static Setting<Boolean> reverse = new Setting<>("Inverse", false);
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:31]: add descriptions ig

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+
+import java.util.LinkedList;
+
+import static org.lwjgl.opengl.GL11.*;
+import static org.lwjgl.opengl.GL11.glPopMatrix;
+
+
+public class LagESPModule extends Module {
+
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:20]: add module instance

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    public static Setting<Float> width = new Setting<>("Width", 0.1F, 2F, 5F, 1)
+            .setDescription("The width of the lines");
+
+    public static Setting<Boolean> reverse = new Setting<>("Inverse", false);
+
+    public static Setting<fade> fadeSetting = new Setting<fade>("Fade", fade.LINEAR);
+
+
+    LinkedList<RubberBand> list = new LinkedList<>();
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:36]: declare as private

Comment (Wolfsurge) [reply src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:36]: and final as well ig

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+            // the difference in time between the creation and now
+            long timeDelta = System.currentTimeMillis() - time;
+            timeDelta = timeDelta == 0 ? 1 : timeDelta;
+
+
+            /*
+            so normally i would use a higher order function for calcuting the interp,
+            but im lazy and nobody will ever look at this. so i wont
+             */
+            if( reverse.getValue()){
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:179]: space after if (and remove the space after the opening parenthesis)

Comment (Wolfsurge) [reply src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:179]: and probably a space after the closing parenthesis

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+                //calculate the difference
+                Vec3d d = from.subtract(to);
+
+                //set the position to be an interpolated value from `from` to `to`
+                if (fadeSetting.getValue() == fade.LINEAR) {
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:185]: use .equals()

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+                    intermediary = to.addVector(
+                            d.x * ((timeDelta) / (fadeSpeed.getValue() * 1000)),
+                            d.y * ((timeDelta) / (fadeSpeed.getValue() * 1000)),
+                            d.z * ((timeDelta) / (fadeSpeed.getValue() * 1000))
+                    );
+                }
+                if (fadeSetting.getValue() == fade.DYNAMIC) {
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:192]: .equals()

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+                            d.y * ((timeDelta / 1000F) * (fadeSpeed.getValue() / 10F) / distance),
+                            d.z * ((timeDelta / 1000F) * (fadeSpeed.getValue() / 10F) / distance)
+                    );
+                }
+            }
+        }
+
+    }
+
+    private enum fade {
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:225]: capital F

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+
+    private class RubberBand {
+        //where the player was initially
+        private final Vec3d from;
+        //where they got rubberbanded to
+        private final Vec3d to;
+        private final long time;
+
+        Vec3d intermediary;
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:158]: declare private

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+                    reverse.getValue() ? 0.05F : 1F
+
+            );
+            glVertex3d(
+                    r.to.x - mc.getRenderManager().viewerPosX,
+                    r.to.y - mc.getRenderManager().viewerPosY,
+                    r.to.z - mc.getRenderManager().viewerPosZ
+            );
+
+            if(System.currentTimeMillis() - r.time >= fadeSpeed.getValue() * 1000){
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:131]: spaces

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+    }
+
+
+    /* stolen from breadcrumbs.. */
+    @Override
+    public void onRender3D() {
+
+
+        // Render positions
+        list.forEach(r -> {
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:66]: personally i would name the parameter something more specific but its not really an issue i guess

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+
+    @SubscribeEvent
+    public void onPacket(PacketEvent event){
+        /*  when we get rubber-banded */
+        if (event.getPacket() instanceof SPacketPlayerPosLook) {
+            SPacketPlayerPosLook p = (SPacketPlayerPosLook) event.getPacket();
+
+            /* register that a rubberband happened */
+
+            list.add(new RubberBand(
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:48]: i would make this all one line, but again, not really an issue

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+import net.minecraft.util.math.Vec3d;
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
+
+
+import java.util.LinkedList;
+
+import static org.lwjgl.opengl.GL11.*;
+import static org.lwjgl.opengl.GL11.glPopMatrix;
+
+
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:18]: `@author` tag is worth adding

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+                            d.y * ((timeDelta / 1000F) * (fadeSpeed.getValue() / 10F) / distance),
+                            d.z * ((timeDelta / 1000F) * (fadeSpeed.getValue() / 10F) / distance)
+                    );
+                }
+            } else {
+                //calculate the difference
+                Vec3d d = to.subtract(from);
+
+                //set the position to be an interpolated value from `from` to `to`
+                if (fadeSetting.getValue() == fade.LINEAR) {
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:205]: .equals()

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+                    intermediary = from.addVector(
+                            d.x * ((timeDelta) / (fadeSpeed.getValue() * 1000)),
+                            d.y * ((timeDelta) / (fadeSpeed.getValue() * 1000)),
+                            d.z * ((timeDelta) / (fadeSpeed.getValue() * 1000))
+                    );
+                }
+                if (fadeSetting.getValue() == fade.DYNAMIC) {
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:212]: .equals()

**Review by Wolfsurge — COMMENTED** [review]
stuff i noticed (linus/aesthetical should probably look at this as well)

**Comment by oragejuice** [comment]
I should have fixed all of that also added some sanity checks

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 16]
```diff
@@ -43,20 +55,29 @@ public void onPacket(PacketEvent event){
         if (event.getPacket() instanceof SPacketPlayerPosLook) {
             SPacketPlayerPosLook p = (SPacketPlayerPosLook) event.getPacket();
 
-            /* register that a rubberband happened */
+            // if we have eaten a chorus recently, then the teleport is likely that of the chorus
+            if(!lastChorus.passedTime(400, Timer.Format.MILLISECONDS)) return;
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:59]: space after if and no inline returns

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 17]
```diff
@@ -176,20 +197,20 @@ public void calculateIntermediary(){
             so normally i would use a higher order function for calcuting the interp,
             but im lazy and nobody will ever look at this. so i wont
              */
-            if( reverse.getValue()){
+            if (reverse.getValue()){
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:200]: space before bracket

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java [hunk 18]
```diff
@@ -176,20 +197,20 @@ public void calculateIntermediary(){
...
 
                 //calculate the difference
                 Vec3d d = from.subtract(to);
 
                 //set the position to be an interpolated value from `from` to `to`
-                if (fadeSetting.getValue() == fade.LINEAR) {
+                if (fadeSetting.getValue() == Fade.LINEAR) {
```
Comment (Wolfsurge) [inline src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:206]: .equals()

**Comment by oragejuice** [comment]
oh ffs i forgot about that :sob:

File: src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java (hunk 18 above)
Comment (oragejuice) [reply src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java:206]: .equals is literally just more overhead to check '=='
    public final boolean equals(Object var1) {
        return this == var1;
    }

Comment (Wolfsurge) [reply src/main/java/cope/cosmos/client/features/modules/visual/LagESPModule.java]: yeah i know but aesthetical told me to use it in one of my PRs (alt manager)

**Comment by Wolfsurge** [comment]
This looks... fine, now. @linustouchtips should I merge this?

**Pull request merged**